### ISSUE TITLE
Verify that  Kafka is running before starting the KafkaTopicConnector

### ISSUE THREAD

**Issue by wbittles** [odpi/egeria#2989]
A number of different issues have been reported which have been caused by Kafka not being available when Egeria attempts to produce or consume events.

These "bring up" issues have been encountered primarily during Egeria cohort initialisation, but issues have  been reported sending and consuming events.  Resulting in important cohort initialisation message being lost.   

In order to avoid handling each of these events individually, with each occurrence requiring it's own retry handler with error and audit  logging code.

In order to avoid these bring up error handlers, this change will delay the start of the Egeria KafkaTopicConnector until the Zookeeper cluster reports that Kafka has initialised enough to have running brokers.

**Comment by mandy-chessell** [comment]
What happens to the calling components?  Do they get an exception?  Are the events buffered?  
Ho do things restart when kafka/zookeeper come back?

**Comment by planetf1** [comment]
would be helpful to link/respond to the previous issues raised, especially if this proposal supercedes the previous discussion/fix

**Comment by wbittles** [comment]
@planetf1 
There have been a number of reported "bring u"p type issues encountered in the past and this fix will handle all of those. If we image there are different stages to bring up
1. Zookeeper bootstrap
issues HOST_NOT_FOUND
2. Zookeeper reachable but cluster not configured yet.
issue TOPIC_NOT_FOUND
3. Zookeeper cluster is up, but no brokers are running.
issue a Kafka Exception (Execution or timeout)

This fix means that all of the above stages must have completed successfully before we complete successfully. If we can't complete the Exception that occurred last will be reported. 
So If initially Egeria couldn't bootstrap to zookeeper , this exception would be stored. If Egeria bootstraps then encounters another issue then it's that last issue that will be reported, until successful.
In Jeff's specific case , I tested the creation of an Egeria Cohort without a running Kafla cluster , and the cohort failed to initialise because of this fix catching the exception, so I'm confident this fix prevents Egeria from encountering the problem Jeff reported as we will either ride out the condition by waiting or failing.

**Comment by wbittles** [comment]
@mandy-chessell The calling components do receive an exception 
Here is an abridged version the start and end of the exception trace.
[stacktrace.txt](https://github.com/odpi/egeria/files/4551834/stacktrace.txt)

In the case of the cohort test, the cohort failed to start, because this fix detected that a Kafka Cluster wasn't running.
In the case of a single server , the first OMAS that tries to connect receives the exception and the server stops.
I've tested starting an Egeria server before starting a Kafka cluster, then started the Kafka Cluster and Egeria rode out the initial errors and seamlessly recovered when the Kafka cluster became available. This ride out period is configurable with a retry count and a sleep time. 

There are no event's generated as this fix fails the cohort creation, this was an attempt to avoid an inconsistent cohort state, as the cohort will need to be created from scratch each time, until successful, which will only happen when Kafka is able to actually run some brokers. This fix prevents the TopicConnector from buffering cohort initialisation events while it's waiting on a connection.

**Comment by planetf1** [comment]
I think this can be closed now - we should raise a new issue if problems still seen after the recent enhancements

**Comment by wbittles** [comment]
@planetf1 agreed
